File: 王道/第12章：动态规划/最大连续子序列/最大子矩阵.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int max_subsequence(const vector<int>& S){
    int n = S.size();
    vector<int> dp(n);
    dp[0] =S[0];
    int cur_max = S[0];
    for (int i=1; i < n;i++){
        dp[i]=max(dp[i-1]+S[i],S[i]);
        cur_max = max(cur_max, dp[i]);
    }
    return cur_max;
}

int main(){
    int n=0;
    while(scanf("%d",&n)!=EOF){
        vector<vector<int>> arr(n,vector<int>(n));
        vector<vector<int>> dp(n+1,vector<int>(n));
        for (int i=0;i<n;i++){
            for (int j=0;j<n;j++){
                scanf("%d",&arr[i][j]);
            }
        }
        for (int i=0;i<n+1;i++){
            for (int j=0;j<n;j++){
                if (i==0){
                    dp[i][j]=0;
                }
                else{
                    dp[i][j]=dp[i-1][j]+arr[i-1][j];
                }
            }
        }
        
        vector<int> temp(n);
        int max_sum=arr[0][0];
        int current_max=arr[0][0];
        for (int top=0;top<n;top++){
            for (int bottom=top+1;bottom<n+1;bottom++){
                for (int j=0;j<n;j++){
                    temp[j]=dp[bottom][j]-dp[top][j];
                }
                current_max=max_subsequence(temp);
                if(current_max>max_sum){
                    max_sum=current_max;
                }
    
            }
        }
        printf("%d\n",max_sum);
        }
        return 0;
    }
    

File: 王道/第12章：动态规划/最大连续子序列/最大子矩阵.py
import sys
inputs=sys.stdin.read().split()
idx=0

def max_subsequence(A:list[int]):
    dp=[0]*len(A)
    dp[0]=A[0]
    for i in range(1,len(A)):
        dp[i]=max(A[i],dp[i-1]+A[i])
    
    return max(dp)
    
    
while idx<len(inputs):
    N=int(inputs[idx])
    idx+=1
    matrix=[]
    for _ in range(N):
        row_list=[]
        for _ in range(N):
            a=int(inputs[idx])
            idx+=1
            row_list.append(a)
        matrix.append(row_list)
        
    #此处可以用到一个动态规划
    dp = [[0] * N for _ in range(N+1)]
    for j in range(0,N):
        for i in range(1,N+1):  #添加了最上面的一行0，为了计算 整个矩阵的和
            dp[i][j]=dp[i-1][j]+matrix[i-1][j]
    
    if N==1:
        print(max_subsequence(matrix[0]))
    else:
        max_num=matrix[0][0]
        for top in range(0,N):
            for bottom in range(top+1,N+1):
                sum_subsequence=[dp[bottom][k]-dp[top][k] for k in range(N)]
                current_max = max_subsequence(sum_subsequence)
                if current_max > max_num:
                    max_num = current_max
        print(max_num)
    